File: API/Model/Basket.cs
using System;
using Microsoft.AspNetCore.Http.Features;

namespace API.Model;

public class Basket
{
    public int Id { get; set; }

    public required string BasketId { get; set; }

    public required List<BasketItem> Items { get; set; } = [];

    public string? ClientSecret { get; set; }

    public string? PaymentIntentId { get; set; }

    public void AddItem(Product product, int quantity)
    {
        if (product == null) ArgumentNullException.ThrowIfNull(product);
        if (quantity <= 0) throw new ArgumentException("Quantity shoud be greater than zero", nameof(quantity));

        var existingItem = FindItem(product.Id);

        if (existingItem == null)
        {
            Items.Add(new BasketItem
            {
                Products = product,
                Quantity = quantity
            });
        }
        else
        {
            existingItem.Quantity += quantity;
        }
    }
    public void RemoveItem(int productid, int quantity)
    {
        if (quantity <= 0)
        {
            throw new ArgumentException("Quantity shoud be greater than zero", nameof(quantity));
        }
        var item = FindItem(productid);
        if (item == null) return;
        item.Quantity -= quantity;

        if (item.Quantity <= 0) Items.Remove(item);

    }

    private BasketItem? FindItem(int productid)
    {
        return Items.FirstOrDefault(x => x.ProductId == productid);
    }
}

File: Client/src/app/models/Basket.ts
import type { IProduct } from "./product"

export type Basket ={
  basketId: string
  items: Item[]
  PaymentIntentId?:string
  clientSecret?:string
}

export class Item {
  constructor( product:IProduct,quantity:number) {
    this.productId=product.id;
    this.brand=product.brand;
    this.name=product.name;
    this.type=product.type;
    this.pictureUrl=product.pictureUrl;
    this.price=product.price;
    this.quantity=quantity;
  }
  productId: number
  name: string
  price: number
  pictureUrl: string
  type: string
  brand: string
  quantity: number
}